src/core/public/http/external_url_service.ts

```diff
@@ -0,0 +1,108 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { IExternalUrlPolicy } from 'src/core/server/types';
+
+import { CoreService } from 'src/core/types';
+import { IExternalUrl } from './types';
+import { InjectedMetadataSetup } from '../injected_metadata';
+import { Sha256 } from '../utils';
+
+interface SetupDeps {
+  location: Pick<Location, 'origin'>;
+  injectedMetadata: InjectedMetadataSetup;
+}
+
+function* getHostHashes(actualHost: string) {
+  yield new Sha256().update(actualHost, 'utf8').digest('hex');
+  let host = actualHost.substr(actualHost.indexOf('.') + 1);
+  while (host) {
+    yield new Sha256().update(host, 'utf8').digest('hex');
+    if (host.indexOf('.') === -1) {
+      break;
+    }
+    host = host.substr(host.indexOf('.') + 1);
+  }
+}
+
+const isHostMatch = (actualHost: string, ruleHostHash: string) => {
+  for (const hash of getHostHashes(actualHost)) {
+    if (hash === ruleHostHash) {
+      return true;
+    }
+  }
+  return false;
+};
```

Comment:
I like what you did here and it's really elegant.

Nit: There's an edge case that this doesn't work for. A URI's `host` component can be 1. an IP literal encapsulated within square brackets, 2. an IPv4 address in dotted-decimal form, or 3. a registered name ([RFC 3986](https://tools.ietf.org/html/rfc3986#section-3.2.2)). The RFC says:

> A registered name intended for lookup in the DNS uses the syntax
> defined in Section 3.5 of [RFC1034] and Section 2.1 of [RFC1123].
> Such a name consists of a sequence of domain labels separated by ".",
> each domain label starting and ending with an alphanumeric character
> and possibly also containing "-" characters.  **The rightmost domain
> label of a fully qualified domain name in DNS may be followed by a
> single "."** and should be if it is necessary to distinguish between
> the complete domain name and some local domain.

In other words: `google.com` is shorthand for `google.com.`, and this appears to hold true for IPv4 addresses as well (at least, from my local testing in a terminal 🙂).

I'm thinking this could account for the edge case without changing your approach:

<details><summary>Click for diff</summary>

```diff
diff --git a/src/core/public/http/external_url_service.test.ts b/src/core/public/http/external_url_service.test.ts
index d5371be253f..ccb6a5a55fc 100644
--- a/src/core/public/http/external_url_service.test.ts
+++ b/src/core/public/http/external_url_service.test.ts
@@ -33,10 +33,15 @@ const setupService = ({
   serverBasePath: string;
   policy: ExternalUrlConfig['policy'];
 }) => {
-  const hashedPolicies = policy.map((entry) => ({
-    ...entry,
-    host: entry.host ? new Sha256().update(entry.host, 'utf8').digest('hex') : undefined,
-  }));
+  const hashedPolicies = policy.map((entry) => {
+    // If the host contains a period (i.e., it is a registered name or IPv4 address), append a `.` if it doesn't already contain one
+    const hostToHash =
+      entry?.host?.indexOf('.') && !entry.host.endsWith('.') ? `${entry.host}.` : entry.host;
+    return {
+      ...entry,
+      host: hostToHash ? new Sha256().update(hostToHash, 'utf8').digest('hex') : undefined,
+    };
+  });
   const injectedMetadata = injectedMetadataServiceMock.createSetupContract();
   injectedMetadata.getExternalUrlConfig.mockReturnValue({ policy: hashedPolicies });
   injectedMetadata.getServerBasePath.mockReturnValue(serverBasePath);
@@ -332,6 +337,25 @@ describe('External Url Service', () => {
         expect(result?.toString()).toEqual(urlCandidate);
       });
 
+      it('allows external urls with a partially matching host and protocol in the allow list when the URL includes the root domain', () => {
+        const { setup } = setupService({
+          location,
+          serverBasePath,
+          policy: [
+            {
+              allow: true,
+              host: 'google.com',
+              protocol: 'https',
+            },
+          ],
+        });
+        const urlCandidate = `https://www.google.com./foo?bar=baz`;
+        const result = setup.validateUrl(urlCandidate);
+
+        expect(result).toBeInstanceOf(URL);
+        expect(result?.toString()).toEqual(urlCandidate);
+      });
+
       it('allows external urls that specify a locally addressable host', () => {
         const { setup } = setupService({
           location,
diff --git a/src/core/public/http/external_url_service.ts b/src/core/public/http/external_url_service.ts
index 05ebe00e7de..7d720615413 100644
--- a/src/core/public/http/external_url_service.ts
+++ b/src/core/public/http/external_url_service.ts
@@ -42,7 +42,10 @@ function* getHostHashes(actualHost: string) {
 }
 
 const isHostMatch = (actualHost: string, ruleHostHash: string) => {
-  for (const hash of getHostHashes(actualHost)) {
+  // If the host contains a period (i.e., it is a registered name or IPv4 address), append a `.` if it doesn't already contain one
+  const hostToHash =
+    actualHost.indexOf('.') && !actualHost.endsWith('.') ? `${actualHost}.` : actualHost;
+  for (const hash of getHostHashes(hostToHash)) {
     if (hash === ruleHostHash) {
       return true;
     }
diff --git a/src/core/server/external_url/external_url_config.ts b/src/core/server/external_url/external_url_config.ts
index ad2a7940ec0..acd4a5f6112 100644
--- a/src/core/server/external_url/external_url_config.ts
+++ b/src/core/server/external_url/external_url_config.ts
@@ -85,9 +85,12 @@ export class ExternalUrlConfig implements IExternalUrlConfig {
   constructor(rawConfig: IExternalUrlConfig) {
     this.policy = rawConfig.policy.map((entry) => {
       if (entry.host) {
+        // If the host contains a period (i.e., it is a registered name or IPv4 address), append a `.` if it doesn't already contain one
+        const hostToHash =
+          entry.host.indexOf('.') && !entry.host.endsWith('.') ? `${entry.host}.` : entry.host;
         return {
           ...entry,
-          host: createSHA256Hash(entry.host),
+          host: createSHA256Hash(hostToHash),
         };
       }
       return entry;
```

</details>

What do you think?

Reply:
Great find! Addressed in [`85c1dfb` (#81234)](https://github.com/elastic/kibana/pull/81234/commits/85c1dfb72c907422b8b1105f2977ab4a77192dfa)